core/include/edm/container.hpp

```diff
@@ -51,9 +84,163 @@ namespace traccc {
+         * methods in this class.
+         */
+        using element_view = container_element<
+            header_t,
+            typename item_vector::value_type
+        >;
+
+        /**
+         * @brief The size type of this container, which is the type by which
+         * its elements are indexed.
+         */
+        using size_type = typename header_vector::size_type;
         /// @}
 
+        /**
+         * We need to assert that the header vector and the outer layer of the
+         * jagged vector have the same size type, so they can be indexed using
+         * the same type.
+         */
+        static_assert(
+            std::is_same_v<
+                typename header_vector::size_type,
+                typename item_vector::size_type
+            >,
+            "Size type for container header and item vectors must be the same."
+        );
+
+        /**
+         * @brief Standard two-argument constructor.
+         *
+         * To enforce the invariant that both vectors must be the same size, we
+         * check this in the constructor. This is also checked in release
+         * builds.
+         */
+        container(
+            header_vector && hv,
+            item_vector && iv
+        )
+            : headers(hv)
+            , items(iv)
+        {
+            if (headers.size() != items.size()) {
+                throw std::logic_error("Header and item length not equal.");
+            }
+        }
+
+        /**
+         * @brief Argument-forwarding constructor.
+         *
+         * This constructor forwards its arguments to both internal vectors.
+         * Obviously, this requires the arguments to be quite general, but this
+         * constructor can be useful when initializing a container for a
+         * specific vecmem memory resource.
+         */
+        template<class... args>
+        container(
+            args&&... a
+        )
+            : headers(a...)
+            , items(a...)
+        {}
+
+        /**
+         * @brief Bounds-checking mutable element accessor.
+         */
+        element_view
+        at(
+            const size_type & i
+        ) {
+            if (i >= size()) {
+                throw std::out_of_range("Index out of range.");
+            }
+
+            return operator[](i);
+        }
+
+        /**
+         * @brief Bounds-checking immutable element accessor.
+         */
+        const element_view
+        at(
+            const size_type & i
+        ) const {
+            if (i >= size()) {
+                throw std::out_of_range("Index out of range.");
+            }
+
+            return operator[](i);
+        }
+
+        /**
+         * @brief Mutable element accessor.
+         */
+        element_view
+        operator[](
+            const size_type & i
+        ) {
+            return {headers.at(i), items.at(i)};
+        }
+
+        /**
+         * @brief Immutable element accessor.
+         */
+        const element_view
```

Comment:
This is not how this should be implemented.

You should rather define a separate `const_element_view` type, like:

```c++
using const_element_view = container_element< const header_t, const item_t >;
```

This would ensure that the caller does not get non-const access to the variables. Since the const-ness of the returned value is lost as soon as the user does something like:

```c++
my_container::element_view el = foo[ i ];
```

Reply:
In which way are `const container_element<A, B>` and `container_element<const A, const B>` not equivalent here?

Reply:
I see your point, but I would propose that we instead add mutable and immutable accessor methods to the `element_view` class. That is, in my opinion, a far more elegant solution than to add these typedefs.

Reply:
😕  This is exactly what we talked about with @scott-snyder in a Core SW meeting about a month ago. When you create such view objects, it's very easy to expose non-const references to vector elements from inside of const accessor functions.

Have another look at the example I gave above. That would expose non-const references from a possibly const container.

Reply:
I assume `foo` is a container here? If that container is const, the only `operator[]` available for it will be the const one, returning a const `element_view`. Then, this `element_view` would only have const accessor methods available, which in turn can only return const references to the contained objects. I fail to see how you would get a non-const reference in this way, can you explain?

Reply:
You don't return a const **reference** from this function. You return a const **value**. As such, my example line shows how the const-ness of that value goes away just by receiving it into a non-const value.

(In general returning a const value of anything from a function has not much of a meaning in C(\+\+)...)

Reply:
Of course. :man_facepalming: